sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/PartitionScopedRegionLevelProgress.java

```diff
@@ -0,0 +1,365 @@
+
+            int regionId = localLsnByRegionEntry.getKey();
+            String normalizedRegionName = RegionNameToRegionIdMap.getRegionName(regionId);
+
+            // the regionId to normalizedRegionName does not exist
+            if (normalizedRegionName.equals(StringUtils.EMPTY)) {
+
+                if (logger.isDebugEnabled()) {
+                    logger.debug("regionId with value - {} which has no known region name has been seen in the vector session token", regionId);
+                }
+
+                request.requestContext.getSessionTokenEvaluationResults().add("Resolving to the global session token since session token provided doesn't have a known region name mapping for region id : " + regionId + ".");
+                return globalSessionToken;
+            }
+
+            // start constructing the session token using the localLsn and globalLsn from lesser preferred regions
+            if (lesserPreferredRegionsPkProbablyRequestedFrom.contains(normalizedRegionName)) {
+                request.requestContext.getSessionTokenEvaluationResults().add("Resolving region specific progress from " + normalizedRegionName);
+                RegionLevelProgress satelliteRegionProgress = this.resolvePartitionKeyRangeIdBasedProgress(partitionKeyRangeId, normalizedRegionName);
+                globalLsn = Math.max(globalLsn, satelliteRegionProgress.maxGlobalLsnSeen);
+                sbPartTwo.append(VectorSessionToken.SegmentSeparator);
+                sbPartTwo.append(regionId);
+                sbPartTwo.append(VectorSessionToken.RegionProgressSeparator);
+                sbPartTwo.append(satelliteRegionProgress.maxLocalLsnSeen);
+            } else {
+                request.requestContext.getSessionTokenEvaluationResults().add("No region specific progress to resolve from " + normalizedRegionName);
+                sbPartTwo.append(VectorSessionToken.SegmentSeparator);
+                sbPartTwo.append(regionId);
+                sbPartTwo.append(VectorSessionToken.RegionProgressSeparator);
+                sbPartTwo.append(-1);
+            }
+        }
+
+        sbPartOne.append(sessionTokenVersion.v);
+        sbPartOne.append(VectorSessionToken.SegmentSeparator);
+        sbPartOne.append(globalLsn);
+
+        Utils.ValueHolder<ISessionToken> resolvedSessionToken = new Utils.ValueHolder<>(null);
+
+        // one additional step of merging base session token / first preferred read region and resolved session token
+        if (SessionTokenHelper.tryParse(sbPartOne.append(sbPartTwo).toString(), resolvedSessionToken)) {
+            return baseSessionToken.merge(resolvedSessionToken.v);
+        }
+
+        request.requestContext.getSessionTokenEvaluationResults().add("Resolving to the global session token since session token from the first preferred region couldn't be merged with region-resolved session token : " + resolvedSessionToken.v.convertToString() + ".");
```

Comment:
I assume you want to have these evaluation results for debugging (to be able to put in CosmosDiagnostics)? - If so, can we avoid this on the hot path (when everything is fine)? Or make it opt-in based on logging level or similar? It is not the end-of-the world since your regression test only shows 2% regression with region-scoped session tokens. But worth at least checking in a CPU profile how much of these 2 percent are spend here?

Reply:
I took a CPU profile of a point read workload (1 million operations with 128 as concurrency) - I see 0.3 to 0.4 percent increase in CPU utilization but nothing that stands out from the above change.